rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md

```diff
@@ -0,0 +1,112 @@
+
+## Internal Proposal
+
+This is roughly the same as https://github.com/timberio/vector/issues/6330#issuecomment-772562955 with some slight tweaks.
+
+A new `emit_log` function will be added to the VRL stdlib:
+
+```text
+emit_log(value: Object)
+```
+
+This function will cause the object passed as value to be emitted at that point and flushed downstream. The emitted log will have its metadata copied from the input event.
+
+Additionally, an `emit_root` (we can work on the naming) config option will be added to the `remap` transform to configure whether `.` is emitted after the transform runs. It will default to `true` to preserve the current behavior but can be set to `false` by users to supress this behavior. Admittedly, I'm not wild about introducing this additional config option, but I'm not seeing another great alternative.
+
+This will be able to be combined with the iteration mechanism that will be introduced [#6031](https://github.com/timberio/vector/issues/6031) to emit an unknown number of events. Naively this might look something like:
+
+```text
+for stooge in .stooges
+   emit(stooge)
+end
+```
+
+In the future we can also add functions for emitting metrics like:
+
+```text
+emit_counter(namespace: String, name: String, timestamp: Timestamp, value: Float, kind: "absolute"|"relative")
+```
+
+I considered having just an `emit_metric()` but it would require users to pass in objects that match exactly the internal representation we have for metrics.
+
+## Doc-level Proposal
+
+The `remap` tranform would gain an extra configuration option:
+
+```text
+emit_root = true/false # default false
+```
+
+When `emit_root` is `true`, the value of `.` will be emitted at the end of the remap program. When `emit_root` is false, the value of `.` will not be emitted. Instead users should use the `emit_log` function to emit.
+
+## Rationale
+
+This enhances remap to support additional transformation use-cases of splitting up an event into multiple events. Without this, users will continue to have to use Lua to acheive this, which introduces a performance bottleneck compared to remap
+
+## Prior Art
+
+- [fluent-plugin-record_splitter](https://github.com/ixixi/fluent-plugin-record_splitter)
+- [Logstash split](https://www.elastic.co/guide/en/logstash/current/plugins-filters-split.html)
+
+These are similar to the `explode` transform PR we have: https://github.com/timberio/vector/pull/6545
+
+## Drawbacks
+
+- It makes the remap transform more complicated to understand and adds an additional configuration option that users need to be aware of
+- Ongoing maintenance burden should be minimal
+
+## Alternatives
+
+### Not emitting . if an `emit_log` is present in the program
+
+We could avoid having an `emit_root` config option on the remap transform by just not emitting automatically if we see an `emit_log` function in the user-provided source. I personally think this would be a bit suprising, but it is an option.
```

Comment:
I was going to suggest this, but I agree it is a potentially surprising behavior. It also may introduce variable behavior if the script does call `emit_log` in some code path.